New module: website_sale_cart_preview

LeartS opened the pull request:
Shows a preview of your cart on mouseover.
Extracted and adapted from the equivalent version 9 feature.

LeartS committed:
New module: website_sale_cart_preview

Shows a popover preview of your cart on mouseover

pedrobaeza commented:
If this is from Odoo, you should extract commit history with git filter-branch and then apply your changes. They need also to be credited.

LeartS commented:
Didn't we already discuss this in https://github.com/OCA/website/pull/53 ?

pedrobaeza commented:
Sorry, I didn't remember the casuistic. It still lacks the credits to Odoo S.A.

LeartS committed:
Add Odoo SA in the credits

oscarolar commented on website_sale_cart_preview/controllers/main.py:
```diff
@@ -0,0 +1,40 @@
+                _order = order.with_context(pricelist=order.pricelist_id.id)
+            values['suggested_products'] = _order._cart_accessories()
+
+        if post.get('type') == 'popover':
```
You are only adding 3 lines of code and you are breaking inheritance mechanism for the method `cart` on website_sale, please use `super` to inherit correctly. 
a example has been posted on [this commit](https://github.com/JayVora-SerpentCS/e-commerce/commit/5067190ce99befcb0272c48cca1add3c945b1f3b), it was the same case, it can be solved the same way.

LeartS replied:
The ironic thing is that [I was the one that showed the way to do it](https://github.com/OCA/e-commerce/pull/41#issuecomment-132981198) to the author on that PR! :laughing:
I don't remember why I didn't do the same thing here, if I forgot or there was a reason, I'll give it a look tomorrow.

oscarolar replied:
Hahaha, shit happens. :smile:

LeartS replied:
So, I looked into it: I couldn't use that method because I needed the `values` the original implementation creates, but I see that I can extract them from the rendered response the super method returns, so I did that. Unfortunately this means that when type = 'popover', 2 templates get rendered. But I don't think it's a big deal.

oscarolar commented on website_sale_cart_preview/static/src/css/website_sale_cart_preview.css:
```diff
@@ -0,0 +1,26 @@
+.mycart-popover .cart_line {
```
why didn't you used .less instead of css?, this can break namespaces on general css.

LeartS replied:
Because I never wrote .less and I never saw an OCA module using it. But with an example I can do it.

oscarolar replied:
About this, I was looking at the OCA builds on runbot and you are right, it is necessary to install website_less module to compile .less files, and is not available, so nevermind.

oscarolar commented on website_sale_cart_preview/static/src/js/website_sale_cart_preview.js:
```diff
@@ -0,0 +1,45 @@
+$(document).ready(function() {
```
using multiple $(document).ready() (Odoo already uses it in several places) it is not always the neatest thing to do.
maybe you can use animationRegistry to add your JS code limited only to the cart DOM selector, an example here https://github.com/odoo/odoo/blob/8.0/addons/mass_mailing/static/src/js/website_mass_mailing.js#L6

LeartS replied:
I couldn't find any documentation whatsoever on this stuff, but I can try. The source seems quite simple.

LeartS replied:
@oscarolar that seems to be specific for snippets, for example (excluding the source code file name) the `start` method accepts and `editable_mode` param. It probably works, but are we sure it's the correct way to do this? Do you know if there are any more general-purpose registries that could be used for this?

oscarolar replied:
There is no doc to support this way of doing this feature, nevertheless its the most efficent way to load this kind of JS files is after browser finishes retrieveing all base JS  and not abuse over the `.ready();` method, this does not mean that you are wrong or I am right, just is not a good practice, its more about a shape rule, anyway thanks for the great contribution, lets merge it.

LeartS committed:
Do not break inheritance of 'cart' controller.

Call the super implementation and then, if needed,
take the values from the rendered response's qcontext.
This means that the qweb template will be rendered 2 times
in some cases, but there is no way around it.

Also change back 'website_sale_order' to 'order' in the qweb
context so that we can return the super rendered response
without modifications.

pedrobaeza commented:
:+1:

LeartS commented:
Rebased on current 8.0

rafaelbn commented:
Tested in runbot :+1: 
Super thanks!
@dgzurita you should take a look in this module

carlosdauden commented:
:+1:

pedrobaeza commented:
Please move it to e-commerce and I'll merge it at the moment.

LeartS commented:
@pedrobaeza what do you mean? We are in e-commerce repo.

pedrobaeza commented:
Sorry, I have mixed PRs, hehe. My bad.

Let's see if @oscarolar say something about your last comment, and I'll merge.

rafaelbn commented:
@oscarolar Please we are waiting for you comments to merge, PR has already 3 +1. Anything to say? Thanks

oscarolar commented:
you can merge this now :+1:

pedrobaeza commented on website_sale_cart_preview/__openerp__.py:
```diff
@@ -0,0 +1,36 @@
+    'website': "http://www.wearemonk.com",
+    'category': 'Website',
+    'version': '8.0.1.0.0',
+    'license': 'AGPL-3',
```
Last comment. I think this GPL-3, not AGPL-3

LeartS committed:
Fix license

LeartS commented:
@pedrobaeza changed the license, sorry for the elapsed time.
